internal/rbd/controllerserver.go

```diff
@@ -262,27 +257,121 @@ func (cs *ControllerServer) CreateVolume(ctx context.Context, req *csi.CreateVol
 	return &csi.CreateVolumeResponse{Volume: volume}, nil
 }
 
-func createBackingImage(ctx context.Context, cr *util.Credentials, rbdVol *rbdVolume, rbdSnap *rbdSnapshot) error {
+func checkFlatten(ctx context.Context, rbdVol *rbdVolume, cr *util.Credentials) error {
+	err := rbdVol.flattenRbdImage(ctx, cr)
+	if err != nil {
+		if _, ok := err.(ErrFlattenInProgress); ok {
+			return status.Error(codes.Aborted, err.Error())
+		}
+		if errDefer := deleteImage(ctx, rbdVol, cr); errDefer != nil {
+			klog.Errorf(util.Log(ctx, "failed to delete rbd image: %s with error: %v"), rbdVol, errDefer)
+			return status.Error(codes.Internal, err.Error())
+		}
+		errDefer := undoVolReservation(ctx, rbdVol, cr)
+		if errDefer != nil {
+			klog.Warningf(util.Log(ctx, "failed undoing reservation of volume: %s (%s)"), rbdVol.RequestName, errDefer)
+		}
+		return status.Error(codes.Internal, err.Error())
+	}
+	return nil
+}
+
+func (cs *ControllerServer) createVolumeFromSnapshot(ctx context.Context, cr *util.Credentials, rbdVol *rbdVolume, snapshotID string) error {
+	rbdSnap := &rbdSnapshot{}
+	if acquired := cs.SnapshotLocks.TryAcquire(snapshotID); !acquired {
+		klog.Infof(util.Log(ctx, util.SnapshotOperationAlreadyExistsFmt), snapshotID)
+		return status.Errorf(codes.Aborted, util.VolumeOperationAlreadyExistsFmt, snapshotID)
+	}
+	defer cs.SnapshotLocks.Release(snapshotID)
+
+	err := genSnapFromSnapID(ctx, rbdSnap, snapshotID, cr)
+	if err != nil {
+		if _, ok := err.(util.ErrPoolNotFound); ok {
+			klog.Errorf(util.Log(ctx, "failed to get backend snapshot for %s: %v"), snapshotID, err)
+			return status.Error(codes.InvalidArgument, err.Error())
+		}
+		return status.Error(codes.Internal, err.Error())
+	}
+
+	// update parent name(rbd image name in snapshot)
+	rbdSnap.RbdImageName = rbdSnap.RbdSnapName
+	// create clone image and delete snapshot
+	err = cloneRbdImageFromSnapshot(ctx, rbdVol, rbdSnap)
+	if err != nil {
+		klog.Errorf(util.Log(ctx, "failed to clone rbd image %s from snapshot %s: %v"), rbdSnap, err)
+		return err
+	}
+
+	err = rbdVol.flattenRbdImage(ctx, cr)
+	if err != nil {
+		return err
+	}
+	klog.V(4).Infof(util.Log(ctx, "create volume %s from snapshot %s"), rbdVol.RequestName, rbdSnap.RbdSnapName)
+	return nil
+}
+
+func (cs *ControllerServer) createBackingImage(ctx context.Context, cr *util.Credentials, rbdVol *rbdVolume, rbdSnap *rbdSnapshot) error {
 	var err error
 
+	var j = new(journal.Connection)
```

Comment:
Why new? It's exceedingly rare to see a call to new used in most go code IMO.

Reply:
>internal/rbd/controllerserver.go:783:2: SA4006: this value of `j` is never used (staticcheck)
	j := &journal.Connection{}

the short variable declaration was giving a static check error for this now, now change back to address of struct.